12/README.md

```diff
@@ -0,0 +1,165 @@
+
+*Tango Server Library* (server kernel library) - library that encapsulates *Tango Server* features and provides *Tango Server API* to a user.
+
+*Tango Server API* (server API) - an API that user uses to develop *Tango Device* functionality.
+
+*Upstream Tango Server/Device* (upstream server) - a *Tango Device* to which updates *client* subscribes for.  
+
+*Polling thread* (polling thread) - a thread within *Tango Server* that polls periodically defined *Tango Device*'s attributes/sometimes commands.
+
+*Tango Publish-Subscriber protocol* (pub-sub) - is a sub-set of *Tango Client-Server protocol* [RFC-?] that defines a relationship between *client* and *server* established via a negotiation phase and lasts till explicitly broken. During this relationship client receives updates published by *upstream server*. Updates are sourced from server's *polling thread* or directly through *server API*. Updates are filtered according to *device*'s configuration.
+
+*Tango Event* (event) - an update sent by *server*.
+
+*Tango Event Callback* (callback) - a piece of code that *client library* executes when an update is received.
+
+## Definitions
+
+This section offers a number of ABNF formal definitions for entities repeatedly used below in this document. 
+
+```ABNF
+EVENT_TYPE = "QUALITY_EVENT" | "INTERFACE_CHANGE" | "PIPE" | "ATTR_CONF_EVENT" |  "CHANGE_EVENT" | "PERIODIC_EVENT" | "ARCHIVE_EVENT" | "USER_EVENT" | "HEARTBEAT"
+
+HEARTBEAT_ENDPOINT = "(upstream server FQDN):port"  
+
+EVENT_ENDPOINT = "(upstream server FQDN):port"
+
+HEARTBEAT_CHANNEL = "(upstream admin device FQDN)/heartbeat" 
+
+EVENT_CHANNEL = "(upstream admin device FQDN)/(attr).EVENT_TYPE[_idl5]" 
+```
+
+## Runtime requirements
+
+Client and server are up and running. Server is reachable from client i.e. may communicate using Request-Reply protocol [RFC-10]. 
+
+## Publisher-Subscriber protocol
+
+![Subscription_1.png](images/Subscription_1.png)
+
+The main goal of the this protocol implementation is to allow client receive events from server:
+
+* Server MUST publish events of EVENT_TYPE
+
+* Server library MUST provide an API to publish events of EVENT_TYPE
+
+* Client SHOULD process events
+
+* Client library SHOULD provide user an API to register user defined event callbacks  
+
+### Negotiation
+
+![Subscription_2.png](images/Subscription_2.png)
+
+Negotiation phase implies that client executes admin's EventSubscriptionChange command:
+
+* Admin MUST implement EventSubscriptionChange command: 
+
+```ABNF
+EventSubscriptionChange = Tango Command
+EventSubscriptionChange_ARGIN-type  = DevVarStringArr 
+EventSubscriptionChange_ARGIN-value = "[ (upstream device), (attribute), "subscribe", EVENT_TYPE]"
+EventSubscriptionChange_ARGOUT-type = DevLong
+EventSubscriptionChange_ARGOUT-value = "(server library release ver)"
+```
+
+**NOTE**: Tango Server Library v9.3.3 force client to explicitly call ZmqEventSubscriptionChange to use ZMQ based implementation [RFC-13]: 
```

Comment:
The note and the ABNF below should go to  #39 
but here should be specify that a specific protocol should use their own set of command on the Admin Device (unfortunately):

```ABNF
XyzEventSubscriptionChange = Tango Command
````
Where Xyz is the code name of the specific P/S protocol.

Reply:
Another question: Are to one to one compatible? ZmqEventSubscriptionChangeArgument-value compatible with EventSubscriptionChange_ARGIN-value

Reply:
Another question: Is the command EventSubscriptionChange only for notifyd?